commons/src/main/scala/org/bitbucket/inkytonik/cooma/Compiler.scala

```diff
@@ -76,24 +76,54 @@ trait Compiler {
 
         def compileTopArg(a : String, t : Expression, e : Expression) : Term = {
 
-            def compileCapArg(n : String) : Term = {
+            def compileCapArg(n : Seq[String]) : Term = {
                 val x = fresh("x")
-                letV(x, prmV(argumentP(nArg), Vector()),
-                    letV(a, prmV(capabilityP(n), Vector(x)),
-                        compileTop(e, nArg + 1)))
+                def aux(n : Seq[String], prev : String) : Term = {
+                    val c0 = fresh("c")
+                    n match {
+                        case hd +: Nil =>
+                            letV(c0, prmV(capabilityP(hd), Vector(x)),
+                                letV(a, prmV(recConcatP(), Vector(prev, c0)),
+                                    compileTop(e, nArg + 1)))
+                        case hd +: (tl @ _ +: _) =>
+                            val c1 = fresh("c")
+                            letV(c0, prmV(capabilityP(hd), Vector(x)),
+                                letV(c1, prmV(recConcatP(), Vector(prev, c0)),
+                                    aux(tl, c1)))
+                        case n =>
```

Comment:
I think here (and the similar one below) it would be clearer to match on Nil, not n, since I think that's the only case that's not already covered.

Also, can you use `sys.error("compileCapArg: unexpected Nil")` instead of just throwing the match error, so the message is a more specific and it follows the style elsewhere?

Reply:
Yes, that's what I originally had but the Scala compiler reported non-exhaustive pattern matching.

Reply:
Ah, yeah this is tricky. I think part of the issue is that `Nil` is an empty list, not an empty `Seq`. So you're not covering as many sequences as you think (e.g. the first case just matches singleton lists). You can avoid that by using `Seq(hd)` as the pattern for that case.

However, I think there's a bigger problem: the compiler's exhaustivity checking basically ignores extractors and `+:` is an extractor. 

My suggestion would be to move from `Seq` to `List` since we don't really need general sequences here. Then you can use `::` instead of `:+`. The former is a constructor, not an extractor, so the exhausitivity checking should work better.

Reply:
Okay I will do that.

`Nil` matches any empty `Seq`, even if it isn't a `List`. (I'm 99% sure of this because I've done this many times before. I almost never use `List` or `Vector` in my own projects, only `Seq`.)

I am a little confused as to why the compiler can't see exhaustiveness in this case. I don't think I've ever seen an exhaustiveness warning from this before:

```
xs match {
  case hd +: tl => ???
  case Nil => ???
}
```

I'll also add the error message. The `Nil` case should be infallible though.

Reply:
Yeah, it's confusing. I assume there's some automatic conversion going on somewhere. I've definitely had previous trouble accidentally using Nil when I really needed an empty general sequence.

They made some significant changes to exhaustivity checking in 2.13.4. I think in the past the checking was more patchy, but less prone to confusion from dodgy warnings.